rust/lsp-lib/src/language_server_client.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation for Language Server Client
+
+use jsonrpc_lite::{Error, Id, JsonRpc, Params};
+use lsp_types::*;
+use serde_json;
+use serde_json::{to_value, Value};
+use std::collections::HashMap;
+use std::io::Write;
+use std::process;
+use types::Callback;
+use url::Url;
+use xi_core::ViewIdentifier;
+
+/// A type to abstract communication with the language server
+pub struct LanguageServerClient {
+    writer: Box<Write + Send>,
+    pending: HashMap<u64, Callback>,
+    next_id: u64,
+    language_id: String,
+    pub is_initialized: bool,
+    pub opened_documents: HashMap<ViewIdentifier, Url>,
+    pub server_capabilities: Option<ServerCapabilities>,
+    pub file_extensions: Vec<String>,
+}
+
+/// Prepare Language Server Protocol style JSON String from
+/// a serde_json object `Value`
+fn prepare_lsp_json(msg: &Value) -> Result<String, serde_json::error::Error> {
+    let request = serde_json::to_string(&msg)?;
+    Ok(format!(
+        "Content-Length: {}\r\n\r\n{}",
+        request.len(),
+        request
+    ))
+}
+
+/// Get numeric id from the request id.
+fn number_from_id(id: Id) -> u64 {
```

Comment:
I think this method is safe. The method is being called only in case of Success and Error message from Language Server as a reply to request. As per the specification, the server _must_ send the same Id as the request id which is number as sent by our client. If we are able to parse the message from client to the JsonRpc Enum as a Success or Error variant, the id (non - None) must be present.

Reply:
okay, sounds good.